aspnetcore/blazor/call-javascript-from-dotnet.md

```diff
@@ -691,6 +669,92 @@ Consider the following guidance when developing code that transfers a large amou
 
+## Unmarshalled JS interop
+
+In Blazor WebAssembly apps when the serialization of many or large .NET objects results in poor app performance, <xref:Microsoft.JSInterop.IJSUnmarshalledObjectReference> represents a reference to an JavaScript object whose functions can be invoked without the overhead of serializing .NET data.
+
+In the following example:
+
+* A [struct](/dotnet/csharp/language-reference/builtin-types/struct) containing a string and an integer is passed unserialized to JavaScript.
+* The JavaScript function (`unmarshalledFunction`) confirms the receipt of the data by checking field values and returning `true` when they match expected values.
+
+> [!NOTE]
+> The following example isn't a typical use case for this scenario because the [struct](/dotnet/csharp/language-reference/builtin-types/struct) passed to JavaScript is small and doesn't result in poor component performance. The example uses a small object merely to demonstrate the concepts for passing unserialized .NET data.
+
+```javascript
+window.returnJSObjectReference = () =>
+{
+  return {
+    dispose: function () {
+      DotNet.disposeJSObjectReference(this);
+    },
+    unmarshalledFunction: function (fields) {
+      const name = Blazor.platform.readStringField(fields, 0);
+      const year = Blazor.platform.readInt32Field(fields, 8);
+
+      return name === "Brigadier Alistair Gordon Lethbridge-Stewart" &&
+        year === 1968;
+    }
+  };
+}
+```
+
+```razor
+@page "/unmarshalled-js-interop"
+@using System.Runtime.InteropServices
+@using Microsoft.JSInterop
+@inject IJSRuntime JS
+
+<h1>Unmarshalled JS interop</h1>
+
+@if (result)
+{
+    <p>JS interop was successful!</p>
+}
+
+<p>
+    <button @onclick="CallJSUnmarshalled">Call JS Unmarshalled</button>
+</p>
+
+<p>
+    <a href="https://www.doctorwho.tv">Doctor Who</a> 
+    is a registered trademark of the <a href="https://www.bbc.com/">BBC</a>.
+</p>
+
+@code {
+    private bool result;
+
+    private void CallJSUnmarshalled()
+    {
+        var unmarshalledRuntime = (IJSUnmarshalledRuntime)JS;
+
+        var jsUnmarshalledReference = unmarshalledRuntime
+            .InvokeUnmarshalled<IJSUnmarshalledObjectReference>(
+                "returnJSObjectReference");
+
+        result = jsUnmarshalledReference.InvokeUnmarshalled<InteropStruct, bool>(
+            "unmarshalledFunction",
+            new InteropStruct
+            {
+                Name = "Brigadier Alistair Gordon Lethbridge-Stewart",
+                Year = 1968,
+            });
+    }
+
+    [StructLayout(LayoutKind.Explicit)]
+    public struct InteropStruct
+    {
+        [FieldOffset(0)]
+        public string Name;
+
+        [FieldOffset(8)]
+        public int Year;
+    }
+}
+```
+
+Integer array types (`int[]` in C#) and strings (`string` in C#) don't match between .NET and JavaScript and require additional code to pass back to .NET from unmarshalled JavaScript functions.
```

Comment:
The `string` one is fairly easy. You could document it with the caveat that the JS APIs are subject to breaking change:

```js
function unmarshalledFunctionReturningString() {
    BINDING.js_string_to_mono_string("Hello");
}
```

which can be read as `jsUnmarshalledReference.InvokeUnmarshalled<string>`. Arrays are a bit more complicated. Some numeric arrays can be converted using `js_typed_array_to_array` (https://github.com/dotnet/aspnetcore/blob/648c15dbe90fb8c113f7c6b4adeb40d9e10494f6/src/Components/Web.JS/src/Platform/WebAssemblyConfigLoader.ts#L15) but it must be a [typed array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Typed_arrays). You could do other data types (such as string arrays), but it's much less straightforward: 

https://github.com/dotnet/aspnetcore/blob/e8dade6436c7359920d726084e4fd80cb54a29e5/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts#L382-L386

That one is an array of bytes, but it could just as well be an array of strings. On the receiving end, I think you're allowed to read arrays as `object[]`